File: hyeonjun/12week/search_a_2D_matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def binary_search(matrix: List[int], target: int) -> int:
            start = 0
            end = len(matrix) - 1
            mid = (start+end) // 2

            while start <= end:
                if matrix[mid] == target:
                    return mid
                elif matrix[mid] < target:
                    start = mid + 1
                else:
                    end = mid - 1
                mid = (start+end)//2

            return mid

        first_column = [row[0] for row in matrix]
        cand_row = matrix[binary_search(first_column, target)]
        return True if target == cand_row[binary_search(cand_row, target)] else False

File: JeongGod/12week/leetcode/SearchA2DMatrix.java
import java.util.*;
class SearchA2DMatrix {
    public boolean searchMatrix(int[][] matrix, int target) {
        for (int i=0; i<matrix.length; i++) {
            // 가능한 열을 찾는다.
            if (matrix[i][0] <= target && target <= matrix[i][matrix[i].length-1]) {
                // target을 찾는다.

                return Arrays.binarySearch(matrix[i], target) >= 0;

            }
        }
        return false;
    }
}